### PR TITLE
USWDS - Colors: Debug color contrast issues.

### LINKED ISSUE #3906: Contrast calculation for identifier inaccurate
When compiling the sass I get an error stating "Neither "primary" nor "base-lightest" can be AA contrast links and hovers on a "base-darkest" background." This is driven by the link settings in _uswds-theme-colors. Using the default works because "white" is ignored, but any changes cause this error. Below are the setting and clearly the difference between 90 and 5 or 30 is more than 50.

$theme-color-base-lightest: "gray-5";
$theme-color-base-darkest: "gray-90";
$theme-identifier-background-color: "base-darkest";

// Links
$theme-link-color: "primary";
$theme-link-visited-color: "violet-70v";
$theme-link-hover-color: "primary-dark";
$theme-link-active-color: "primary-darker";
$theme-link-reverse-color: "white";
$theme-link-reverse-hover-color: "gold-30";
$theme-link-reverse-active-color: "white";

Using Visual Studio Code and Live Sass Compiler.

I believe the issue may be that set-link-from-bg() doesn't actually test the fallback colors.

### PR DESCRIPTION
**Pull request by mejiaj**
## Description

Fixes #3906 and #3926. 

- Adds [sass-true](https://github.com/oddbird/true) for Mocha-based Sass unit testing (`npm run test:sass`)
- Add standalone test script for selected Sass functions
- Add options `$context` data to color-picker functions for clearer error messaging
- Tried to address many edge cases, and test for them
- Improves performance of color calculations
- Improves error messaging
- Demotes most color errors to warnings, with predictable fallbacks
- Continues to throw errors for `transparent` color calculations
- Adds special error handling to output some error messages as strings for testing purposes
- Adds new helper functions for repeatable, single-purpose calculations:
  - `is-accessible-magic-number()`: Determines if two grades achieve a specified WCAG level
  - `color-token-family()`: Outputs the family of a color token or split
  - `color-token-grade()`: Outputs the grade (system or theme grade) of a color token or split. Note that this is different from `calculate-grade()` which will output the absolute system grade for any color.
  - `color-token-type()`: Outputs the type ("system" or "theme") of a color token
  - `color-token-variant()`: Outputs the variant ("vivid") of a color token or split
  - `magic-number()`: Outputs the magic number between two grades
  - `next-token()`: Gets the next color token "darker" or "lighter" from a specified color

### CODE REVIEW

File: src/stylesheets/core/_variables.scss [hunk 1]
```diff
@@ -406,7 +406,7 @@ $assignments-theme-color: (
   "base-dark": $theme-color-base-dark,
   "base-darker": $theme-color-base-darker,
   "base-darkest": $theme-color-base-darkest,
-  "ink": $theme-color-base-darkest,
+  "ink": $theme-color-base-ink,
```
Comment (mejiaj) [inline src/stylesheets/core/_variables.scss:409]: Discovered while debugging [this comment #3906](https://github.com/uswds/uswds/issues/3906#issuecomment-758776179).

Given a custom Ink value
```scss
// Base colors
$theme-color-base-ink: #141414;
```

It was causing the [custom grades](https://github.com/uswds/uswds/pull/3949/files#diff-9bf0d9df140d95fba8158d3f6fd0876a22565f243cf9a2fcae81804304b59eceR1979-R1980) to return `ink-darker`, which doesn't exist.

File: src/stylesheets/theme/_uswds-theme-components.scss [hunk 2]
```diff
@@ -100,6 +100,7 @@ $theme-identifier-font-family: "ui";
 $theme-identifier-background-color: "base-darkest";
 $theme-identifier-identity-domain-color: "base-light";
 $theme-identifier-max-width: "desktop";
+$theme-identifier-primary-link-color: "base-lighter";
```
Comment (mejiaj) [inline src/stylesheets/theme/_uswds-theme-components.scss:103]: Created this new setting for the parent agency link so users can more easily override.

File: src/stylesheets/components/_identifier.scss [hunk 3]
```diff
@@ -5,10 +5,19 @@ $identifier-link-gap: 1.5;
...
+    $theme-color-base-darker
+  );
+}
+
 @mixin identifier-secondary-link {
   @include set-link-from-bg(
     $theme-identifier-background-color,
-    $theme-identifier-secondary-link-color
+    $theme-identifier-secondary-link-color,
+    $theme-color-base-dark
```
Comment (mejiaj) [inline src/stylesheets/components/_identifier.scss:20]: Both primary and secondary link mixin colors for `set-link-from-bg()` have been updated for clarity. By default it's a dark theme and we want the links to be base-light family.

On light themes they should be the opposite, instead of the primary link color.

File: src/stylesheets/core/_functions.scss [hunk 4]
```diff
@@ -1776,185 +2050,139 @@ system z-index
...
-            nth($custom-split, 3),
-            "default"
-          );
-        } @else {
-          $custom-grade: if(
-            length($custom-split) == 2,
-            nth($custom-split, 2),
-            "default"
-          );
-        }
```
Comment (mejiaj) [inline src/stylesheets/core/_functions.scss:1837]: Separated the custom color code into their own functions w/ documentation so it's easier to understand and debug in the future.

- get-custom-color-shades()
  - is-custom-color()
  - get-custom-color-split()
  - get-custom-color-family()
  - get-custom-color-grade()
  - get-custom-color-darker-grade()
  - get-custom-color-lighter-grade()

File: src/stylesheets/core/_functions.scss [hunk 5]
```diff
@@ -22,6 +22,43 @@ General-purpose functions
...
+
+/*
+----------------------------------------
+error-not-token()
+----------------------------------------
+Returns a common not-a-token error.
+----------------------------------------
+*/
+
+@function error-not-token($token, $type, $valid-token-map: false) {
```
Comment (thisisdano) [inline src/stylesheets/core/_functions.scss:51]: Tried to capture this common error in a repeatable function

File: src/stylesheets/core/_functions.scss [hunk 6]
```diff
@@ -1698,263 +1814,553 @@ system z-index
...
-        @warn 'Tokens with grades less than 10 (including "white") aren\'t valid link color tokens, since they have no lighter hover states.';
-      } @else if $link-grade == 100 {
-        @warn '"black" isn\'t a valid link color token, since it has no darker hover state.';
-      }
+color-token-variant(("red", 50, "vivid"))
+> "red"
+----------------------------------------
+*/
+
+@function color-token-family($color-token) {
```
Comment (thisisdano) [inline src/stylesheets/core/_functions.scss:2101]: This (and `color-token-family()` and `color-token-grade()`) can accept a token or a split. Split is faster!

**Review by thisisdano — APPROVED** [review]
